### PR TITLE
Fix page retained size by accounting slice base object once for a page

### PR DESCRIPTION
**Pull request by singcha**
Test plan - 
Tested with Presto on Spark failing broadcast queries & Presto small & large queries in production

```
== RELEASE NOTES ==

General Changes
* Fix memory overcounting for ``VariableWidthBlocks`` that could cause queries to fail with exceeded memory limit errors

### CODE REVIEW

File: presto-common/src/main/java/com/facebook/presto/common/Page.java [hunk 1]
```diff
@@ -418,12 +422,17 @@ public Page dropColumn(int channelIndex)
 
     private long updateRetainedSize()
     {
-        long retainedSizeInBytes = INSTANCE_SIZE + sizeOf(blocks);
+        AtomicLong retainedSizeInBytes = new AtomicLong(INSTANCE_SIZE + sizeOf(blocks));
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/Page.java:425]: Are we using `AtomicLong` just for mutability?

File: presto-common/src/main/java/com/facebook/presto/common/Page.java [hunk 2]
```diff
@@ -418,12 +422,17 @@ public Page dropColumn(int channelIndex)
...
+        Set<Object> referenceSet = Collections.newSetFromMap(new IdentityHashMap<>());
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/Page.java:426]: Static import

```suggestion
        Set<Object> referenceSet = newSetFromMap(new IdentityHashMap<>());
```

File: presto-common/src/main/java/com/facebook/presto/common/Page.java [hunk 3]
```diff
@@ -418,12 +422,17 @@ public Page dropColumn(int channelIndex)
 
     private long updateRetainedSize()
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/Page.java:423]: Might be worthwhile to add a dedicated test case additionally in `TestPage`?

File: presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlock.java [hunk 4]
```diff
@@ -164,7 +164,15 @@ public long getRetainedSizeInBytes()
     @Override
     public void retainedBytesForEachPart(ObjLongConsumer<Object> consumer)
     {
-        consumer.accept(slice, slice.getRetainedSize());
+        // Slices with shared base objects, common scenario when a block is deserialized
```
Comment (tdcmeehan) [inline presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlock.java:167]: I would reword this a bit.  Explain the problem (for deserialized pages we can have many views on top of an underlying shared Slice object), how this solves it (by slicing up the view reference from the underlying shared byte array, which will not be double counted thanks to reference counting).

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
Minor comments

File: presto-main/src/test/java/com/facebook/presto/execution/buffer/TestPagesSerde.java [hunk 5]
```diff
@@ -107,6 +110,35 @@ public void testVarcharSerializedSize()
...
+                variableWidthBlockBuilder1.build(),
+                variableWidthBlockBuilder2.build(),
+                bigintBlockBuilder.build())
+                .compact();
+        Page deserializedPage = serde.deserialize(serde.serialize(compactPage));
+
+        double expectedMaxSize = compactPage.getRetainedSizeInBytes() * 1.2; // 120%
+        double actualSize = deserializedPage.getRetainedSizeInBytes();
+        assertTrue(actualSize < expectedMaxSize,
+                "Expected round trip size difference less than 20% of original page");
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/execution/buffer/TestPagesSerde.java:139]: put into the same line

**Review by highker — APPROVED** [review]
LGTM. Would be good if @rschlussel can help to take another look.

File: presto-common/src/main/java/com/facebook/presto/common/Page.java (hunk 1 above)
Comment (highker) [reply presto-common/src/main/java/com/facebook/presto/common/Page.java:425]: I think it's for lambda reference

Comment (singcha) [reply presto-common/src/main/java/com/facebook/presto/common/Page.java:425]: yes, its for lambda reference

File: presto-common/src/main/java/com/facebook/presto/common/Page.java (hunk 3 above)
Comment (singcha) [reply presto-common/src/main/java/com/facebook/presto/common/Page.java:423]: Gets tricky to handcraft. Let me try doing it. Another option is to use serde flow, but then it will be similar to existing test

File: presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlock.java [hunk 6]
```diff
@@ -164,7 +164,15 @@ public long getRetainedSizeInBytes()
...
+        // Account for slice memory & underlying byte array separately
+        if (slice.getBase() != null && slice.hasByteArray()) {
+            consumer.accept(slice, Slices.EMPTY_SLICE.getRetainedSize());
```
Comment (rschlussel) [inline presto-common/src/main/java/com/facebook/presto/common/block/VariableWidthBlock.java:170]: static import EMPTY_SLICE

**Comment by rschlussel** [comment]
Also, looks like some of the code was taken from https://github.com/prestodb/presto/pull/10633.  you should give attribution to the author in the commit message.

**Review by rschlussel — APPROVED** [review]
